### DISCUSSION TITLE
Identificador unico (PK SICO/OREPLAY)

### DISCUSSION THREAD

**Discussion by freefri** [oreplay/oreplay#7] [General]
Respecto al identificador unico (entiendo que estamos hablando de un primary key en el software del SICO nuevo). Para mi la respuesta suele ser: esta clase de campos unicos de identificacion genérica, deberían estar definidos por el sistema software, y no deberian ser compartidos con otro campo (por muy unico que este otro campo pueda ser), es decir, no deberías usar un email como clave primaria, no deberias usar un DNI como clave primaria, no deberias usar un numero de licencia como clave primaria. Otro tema distinto es como maneja el OE sus campos únicos (ahí si podriamos usar un numero de licencia o un dorsal), y lo mismo en sitiming, que podemos usar un membership number (ya que los dorsales suelen ser por corredor)

En sico, por supuesto deberiamos tener los campos ID (unico), licencia (opcional), dorsal (opcional) y sportident (opcional)

- Entonces el campo licencia no podemos usarlo como campo único porque corredores sin licencia lo tendrán vacio
- el campo dorsal puede cambiar de una prueba a otra, asi que tampoco deberia ser campo unico
- DNI podria ser un problema con niños pequeños o extranjeros, asi que no deberia ser un campo unico
- Sportident (chip number) puede cambiar de una prueba a otra

Documentacion de OE https://sportsoftware.de/download/v11/OE2010_Handbook_Eng.pdf (ingles)

Documentacion de OE https://www.sportsoftware.de/download/v11/OE2010_Handbook_Esp.pdf (español)

- Dorsal (Start number) Si esta columna es visible en el modo Inserción, se mostrará con el siguiente dorsal disponible
- Id. de inscripción (Entry Id) Servicios externos de inscripción web pueden dar sus propios identificadores de inscripción para permitir una identificación entre el sistema de inscripciones y OE2010. Este campo puede ser transferido importando archivos siguiendo el formato XML de IOF. Normalmente no se debe modificar este valor

![WhatsApp Image 2023-12-05 at 3 43 02 PM](https://github.com/oreplay/oreplay/assets/2754631/d32498fa-d5cd-47cb-b14e-cfd933d6b547)
(aqui hay una mala traduccion en castellano en el manual de oe)

me parece claro que el ID que nos de SICO deberiamos usarlo en OE como entry id (Id. de inscripcion) o como database id (identificador de la base de datos). Con esto podriamos mantener la licencia en Text3, y el dorsal por separado

**Comment by freefri** [comment]
![image](https://github.com/oreplay/oreplay/assets/2754631/3473297a-ebbe-4c76-b307-ccc7faca61ae)

![image](https://github.com/oreplay/oreplay/assets/2754631/9e9dc11c-0e19-431d-8334-b83e3c95403c)

sólo se desbloquea Id de inscripcion si pones XML (ver el manual)

O hacemos que sico solo trabaje con XML... Que yo diría que no es buena opción (porque no permitiria trabajar y editar desde excel)... O usamos **id de la base de datos** como campo primary key

**Comment by freefri** [comment]
Podríamos hacer un mixto para identificar usuarios. 
- Está claro que 2 personas distintas no pueden tener la misma licencia, ni el mismo DNI. Pero puede ser que no conozcan su licencia o si DNI. 
- Aún así podemos asumir que si nos dan una licencia o un DNI, tenemos la vinculación perfecta. 
- Para la gente que no sabe o no tiene uno de esos 2, podríamos usar su dorsal de 4 dígitos, siempre y cuando asumamos que en una carrera no se van a usar dorsales de 4 dígitos que no sean dorsales oficiales Fedo. Ahí todo el mundo tiene que saber su dorsal.
- La tarjeta SI no la usaría nunca porque es muy intercambiable. 
- Esto nos deja libre el grupo de gente que tiene licencia de día y no tiene DNI, estos no los podemos controlar, y generarán resultados sin enlace en sico

**Comment by freefri** [comment]
- Sico deberia tener un PK (primary key) interna
  - No relacionado con licencia, dni, o dorsal [<= 6 digitos numericos porque oe no admite mas])
  - La PK (si no fuera posible usar la PK para ello, Sico deberia proveer un campo nuevo con estas restricciones) la usaremos como "id de base de datos" en OE (hasta 10 char alfanum [a-z0-9]{10} )
- Oreplay usara la PK de Sico como FK (foreign key) externa Fedo
- Asumimos que SiTiming y SportSoftware van a adaptar su ultima version de software para subir datos a oreplay de forma automatica pero debemos dar soporte a versiones anteriores

**Comment by sergiogpajares** [comment]
# 08/12/2023 meeting

**Primary Keys (PK) to be used**

## PK to be used in FEDO SICO

SICO should've its own UUID to identify its users as a PK. That PK would be used later on by o-replay to match it's runners and users to SICO's ones.

This identifier, should match the following requisites:

- The identifier would be imported as `DatabaseID` in OE.
- Type would be `alphanumeric` of equal or less than 10 characters as this is required by both OE and SI Timing.
- It shouldn't be match to any other field, such as BIB number, license or DNI.

If it's not possible to use a PK for data exchange, then a new field should be provided with the same restrictions.

Therefore, attributes should look as the following table

| Type | Attribute  | COMMENTS          |
|------|------------|-------------------|
|  PK  | UniqueID   | Unique mandatory  |
|      | DNI        | Unique optional   |
|      | BIB number | Unique mandatory  |
|      | License    | Unique optional  |
|      | ...        |                   |

This unique attributes could be used to identify runners as well if required. Specially, when a runner doesn't know its UniqueID and it's necessary to look it up by the organizer/timekeeper.

**Notes:**

- SI Timing allows more than one membership number.
- BIB number *(dorsal)* is an `uint` up to 6 digits in OE and `alphanumeric` up to 10 characters in SI Timing.

### Getting IDs from OE software

Support for old OE2010 should be provided. It can't be retrieved from current IOF xml implementation. An `csv` parser for that reason would be provided.

## PK to be used in O-replay

O-replay would have its own PK that would be a proper UUID. Additional field would be included to match external federations/services.

| Type | Attribute     | COMMENTS          |
|------|---------------|-------------------|
|  PK  | UniqueID      | Unique mandatory  |
|  FK  | Federation    | Optional          |
|      | Federation ID | Optional          |
|      | OEventor ID   | Unique optional   |
|      | ...           |                   |

When events are created, an option can be chosen to decide if the federation ID is used by oreplay to match runners to users.  It can be chosen to ignore those identifiers.

When matching is not automatically done, user would be available to identify their selfs. Javi's model can be preserved in this case.

### Automatic matching

An order list of criteria to identify would be used.
